test/test_upload.py

```diff
@@ -195,3 +198,143 @@ def test_compile_and_upload_combo_with_custom_build_path(run_command, data_dir,
         assert f"Compile {sketch_name} for {board.fqbn} successful" in traces
         assert f"Upload {sketch_path} on {board.fqbn} started" in traces
         assert "Upload successful" in traces
+
+
+def test_compile_and_upload_combo_sketch_with_pde_extension(run_command, data_dir, detected_boards, wait_for_board):
+    assert run_command("update")
+
+    sketch_name = "CompileAndUploadPdeSketch"
+    sketch_path = Path(data_dir, sketch_name)
+
+    # Create a test sketch
+    assert run_command(f"sketch new {sketch_path}")
+
+    # Renames sketch file to pde
+    sketch_file = Path(sketch_path, f"{sketch_name}.ino").rename(sketch_path / f"{sketch_name}.pde")
+
+    for board in detected_boards:
+        # Install core
+        core = ":".join(board.fqbn.split(":")[:2])
+        assert run_command(f"core install {core}")
+
+        # Build sketch and upload from folder
+        wait_for_board()
+        res = run_command(f"compile --clean -b {board.fqbn} -u -p {board.address} {sketch_path}")
+        assert res.ok
+        assert "Sketches with .pde extension are deprecated, please rename the following files to .ino" in res.stderr
+        assert str(sketch_file) in res.stderr
+
+        # Build sketch and upload from file
+        wait_for_board()
+        res = run_command(f"compile --clean -b {board.fqbn} -u -p {board.address} {sketch_file}")
+        assert res.ok
+        assert "Sketches with .pde extension are deprecated, please rename the following files to .ino" in res.stderr
+        assert str(sketch_file) in res.stderr
+
+
+def test_upload_sketch_with_pde_extension(run_command, data_dir, detected_boards, wait_for_board):
+    assert run_command("update")
+
+    sketch_name = "UploadPdeSketch"
+    sketch_path = Path(data_dir, sketch_name)
+
+    # Create a test sketch
+    assert run_command(f"sketch new {sketch_path}")
+
+    # Renames sketch file to pde
+    sketch_file = Path(sketch_path, f"{sketch_name}.ino").rename(sketch_path / f"{sketch_name}.pde")
+
+    for board in detected_boards:
+        # Install core
+        core = ":".join(board.fqbn.split(":")[:2])
+        assert run_command(f"core install {core}")
+
+        # Compile sketch first
+        assert run_command(f"compile --clean -b {board.fqbn} {sketch_path}")
+
+        # Upload from sketch folder
+        wait_for_board()
+        assert run_command(f"upload -b {board.fqbn} -p {board.address} {sketch_path}")
+
+        # Upload from sketch file
+        wait_for_board()
+        assert run_command(f"upload -b {board.fqbn} -p {board.address} {sketch_file}")
+
+        # Upload from build folder
+        sketch_path_md5 = hashlib.md5(bytes(sketch_path)).hexdigest().upper()
+        build_dir = Path(tempfile.gettempdir(), f"arduino-sketch-{sketch_path_md5}")
```

Comment:
You should not compute this value here but get it from `compile --format json` output.